### PR TITLE
New (preview) Web UI scaffolding

### PR DESCRIPTION
**Pull request by koszti**
## Description

**================================================**
**=== This is the re-opened version of the accidentally closed https://github.com/trinodb/trino/pull/22695 ===**
**================================================**

This scaffolding PR aims to modernize the web UI for Trino, incorporating up-to-date tools and a contemporary look and feel.

Partially addressing and proposing technical solutions for https://github.com/trinodb/trino/issues/22697

## Additional context and related issues

Currently, the Trino web UI uses outdated components like React 16, jQuery, and Bootstrap 3. Updating the UI to use more recent tools and frameworks will provide a fresher appearance and improved functionality. For instance, the Trino gateway utilizes [Vite](https://vitejs.dev/) with modern dependencies, suggesting that the Trino web UI could benefit from adopting a similar approach for consistency.

Original slack thread: https://trinodb.slack.com/archives/CKCEWGYT0/p1720904086284609

**IMPORTANT**: This PR does not disrupt the current web app. Instead, it introduces a new web app scaffolding accessible at a new endpoint  (`/ui/preview`), which is visible only in development mode. This approach allows for further development and testing and can switch to the new web app once migrating of all functionalities is completed.

**This PR Utilizes the Following Tools and Development Features:**
* Using [vite.js](https://vitejs.dev/) as the primary frontend tool. Similar to trino-gateway.
* **Hot Module Replacement (HMR)** for instant and precise updates without reloading the page or losing the application state
* Proxying API requests from the development UI to the Trino development server
* Eliminating the need to rebuild the Java project after packaging, unlike the current web app

**What's not included:**
* A React component library. We can decide on one later, with options like [Semi Design](https://semi.design/en-US/) (used in Trino Gateway) or [Ant Design](https://ant.design/), among others.

### To Develop the preview webapp locally

Run the `PreviewUiQueryRunner` class. This will start a minimalistic development server configured with the Preview UI.

Once the dev server is running start the preview webapp in dev mode:
```
$ cd core/trino-main/src/main/resources/webapp-preview
$ npm install && npm run dev

  VITE v5.3.4  ready in 108 ms

  ➜  Local:   http://localhost:5173/ui/preview
  ➜  Network: use --host to expose
  ➜  press h + enter to show help
```

### To Build and package the preview webapp

Build steps are fully automated using Maven, eliminating the need to commit generated JS files to the git repository as was done with the current web app.

The preview web app is disabled by default and enabled only in the full development server by setting the `web-ui.preview.enabled=true` property. This ensures that the preview web app is available only to developers and not to production end users. Once all functionalities are migrated from the old web app to the new one, we can swap them, making the new web app the default for end users.

If the preview webapp is enabled then it's running in parallel to the current one and accessible at `/ui/preview`:

image

## Release notes

(x) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

**Comment by wendigo** [comment]
You've inspired me to actually build the existing UI in a similar way so we don't keep checked in dist directory anymore: https://github.com/trinodb/trino/pull/22785.

**Comment by koszti** [comment]
@mosabua added an initial list of tasks that needs to be completed to the parent issue at https://github.com/trinodb/trino/issues/22697#issuecomment-2248985124. We can review and can link future PRs to it.

**Comment by wendigo** [comment]
@koszti can you rebase this PR? Some shared plugins are already added to the poms. This PR should be much smaller

File: core/trino-main/pom.xml [hunk 1]
```diff
@@ -535,6 +535,11 @@
         <resources>
             <resource>
                 <directory>src/main/resources</directory>
+                <includes>
```
Comment (wendigo) [inline core/trino-main/pom.xml:538]: please drop it, and exclude instead to exclude sources

File: core/trino-main/pom.xml [hunk 2]
```diff
@@ -594,6 +599,41 @@
...
                     </execution>
+                    <execution>
+                        <id>install node and npm (webapp-preview)</id>
+                        <goals>
+                            <goal>install-node-and-npm</goal>
+                        </goals>
+                        <phase>generate-resources</phase>
+                        <configuration>
+                            <workingDirectory>src/main/resources/webapp-preview</workingDirectory>
+                            <installDirectory>src/main/resources/webapp-preview/dist</installDirectory>
```
Comment (wendigo) [inline core/trino-main/pom.xml:610]: can we install dependencies and build in a single logical step?

**Review by wendigo — APPROVED** [review]
LGTM % includes vs excludes

File: core/trino-main/pom.xml (hunk 2 above)
Comment (koszti) [reply core/trino-main/pom.xml:608]: done. Added a new npm command to do things in one go

File: core/trino-main/pom.xml (hunk 1 above)
Comment (koszti) [reply core/trino-main/pom.xml]: done. It's using exclude only

**Comment by wendigo** [comment]
Merged. Thanks @koszti 

cc @mosabua

**Comment by wendigo** [comment]
I've verified that the reproducible builds are not broken

**Pull request merged**